gobblin-service/src/main/java/gobblin/service/modules/orchestration/Orchestrator.java

```diff
@@ -169,45 +116,59 @@ public void onDeleteSpec(URI deletedSpecURI, String deletedSpecVersion) {
-          deletedSpecURI));
-    }
-
-    try {
-      if (flowCatalog.isPresent()) {
-        flowCatalog.get().getSpec(deletedSpecURI);
-        // If deleted Spec is of type Flow, remove it from all SpecExecutorInstances
-        // (SpecExecutorInstances are expected to no-op if they do not have the URI scheduled on them)
-        if (topologyCatalog.isPresent()) {
-          Collection<TopologySpec> specs = specCompiler.getTopologySpecMap().values();
-          for (TopologySpec topologySpec : specs) {
-            try {
-              // Note: At the moment, the URI of Flow and Job is expected to remain the same.
-              // .. We will need to handle 1 Flow URI to n Job URI compilation when we introduce it.
-              // .. Hence at the moment Flow URI is being deleted on all SpecExecutorInstances.
-              topologySpec.getSpecExecutorInstanceProducer().deleteSpec(deletedSpecURI);
-            } catch (RuntimeException e) {
-              _log.warn(String.format("Did not delete Spec with URI: %s and Version: %s on SpecExecutorInstance: %s",
-                  deletedSpecURI, deletedSpecVersion, topologySpec.getSpecExecutorInstanceProducer().getUri()), e);
-            }
-          }
-        }
       }
-    } catch (SpecNotFoundException e) {
-      _log.info(String.format("Spec with URI: %s was not found in TopologyCatalog, going to try FlowCatalog",
-          deletedSpecURI));
+    } catch (RuntimeException e) {
+      _log.info(String.format("Spec with URI: %s was not found in TopologyCatalog", deletedSpecURI), e);
     }
   }
 
   /** {@inheritDoc} */
   @Override
   public void onUpdateSpec(Spec updatedSpec) {
     _log.info("Spec changed: " + updatedSpec);
-    throw new UnsupportedOperationException();
+
+    if (!(updatedSpec instanceof TopologySpec)) {
+      return;
+    }
+
+    try {
+      onDeleteSpec(updatedSpec.getUri(), updatedSpec.getVersion());
+    } catch (Exception e) {
+      _log.error("Failed to update Spec: " + updatedSpec, e);
+    }
+    try {
+      onAddSpec(updatedSpec);
+    } catch (Exception e) {
+      _log.error("Failed to update Spec: " + updatedSpec, e);
+    }
+  }
+
+  public void orchestrate(Spec spec) throws Exception {
+    if (spec instanceof FlowSpec) {
+      Map<Spec, SpecExecutorInstanceProducer> specExecutorInstanceMap = specCompiler.compileFlow(spec);
+
+      if (specExecutorInstanceMap.isEmpty()) {
+        _log.warn("Cannot determine an executor to run on for Spec: " + spec);
+        return;
+      }
+
+      // Schedule all compiled JobSpecs on their respective Executor
+      for (Map.Entry<Spec, SpecExecutorInstanceProducer> specsToExecute : specExecutorInstanceMap.entrySet()) {
+        // Run this spec on selected executor
+        SpecExecutorInstanceProducer producer = null;
+        try {
+          producer = specsToExecute.getValue();
+          Spec jobSpec = specsToExecute.getKey();
+          producer.addSpec(jobSpec);
+        } catch(Exception e){
```

Comment:
producer.addSpec does not throw a typed exception, so catching generic Exception to log it and not crash the thread.